i2c/oled_i2c/img_to_array.py

```diff
@@ -53,27 +51,30 @@
 # our goal is to divide the image into 8-pixel high pages
 # and turn a pixel column into one byte, eg for one page:
 # 0 1 0 ....
-# 1 0 0 
-# 1 1 1 
-# 0 0 1 
-# 1 1 0 
-# 0 1 0 
-# 1 1 1 
+# 1 0 0
+# 1 1 1
+# 0 0 1
+# 1 1 0
+# 0 1 0
+# 1 1 1
 # 0 0 1 ....
 
 # we get 0x6A, 0xAE, 0x33 ... and so on
 # as `pixels` is flattened, each bit in a column is IMG_WIDTH apart from the next
 
 buffer = []
-for i in range(IMG_HEIGHT // OLED_PAGE_HEIGHT):
-    start_index = i*IMG_WIDTH*OLED_PAGE_HEIGHT
-    for j in range(IMG_WIDTH):
+for i in range(img_height // OLED_PAGE_HEIGHT):
+    start_index = i*img_width*OLED_PAGE_HEIGHT
+    for j in range(img_width):
         out_byte = 0
-        for k in range(8):
-            out_byte |= pixels[k*IMG_WIDTH + start_index + j] << k
+        for k in range(OLED_PAGE_HEIGHT):
+            out_byte |= pixels[k*img_width + start_index + j] << k
         buffer.append(f'{out_byte:#04x}')
 
 buffer = ", ".join(buffer)
-buffer_hex = f"static uint8_t[] = {{{buffer}}}"
+buffer_hex = f'static uint8_t {img_name}[] = {{{buffer}}}\n'
```

Comment:
If you were feeling fancy, you could insert a regex-check to check that there were no characters in `img_name` that aren't valid in a C variable-name :slightly_smiling_face:  (spaces, hyphens, etc.)
For extra fanciness you could automatically convert invalid chars to underscores, so that "my awesome-image!.bmp" creates `my awesome-image!.h` containing `static uint8_t my_awesome_image_[]`.

An alternative approach would be to always create the array with the same name, e.g. `static uint8_t image_data[]`.

Reply:
Yeah, lol, I was hoping to outsource as much of that as possible to `pathlib` but I'm happy to give it a go? The only reason I've used the filename is that other examples that use images in header files all have matching names (like the pio LCD one)

Reply:
I dunno how familiar you are with [regexes](https://docs.python.org/3/library/re.html), but `array_name = re.sub(r'[^\w]', '_', img_name)` is probably what you want? (Replace anything NOT in `a-zA-Z0-9_` with an underscore)